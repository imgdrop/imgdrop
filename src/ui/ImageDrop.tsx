// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

import { Link, createStyles, makeStyles } from '@material-ui/core';
import { Panorama } from '@material-ui/icons';
import React from 'react';
import { useDropzone } from 'react-dropzone';

const useStyles = makeStyles((theme) =>
   createStyles({
      root: {
         width: '100vw',
         height: '100vh',
         padding: theme.spacing(2),
         '&:focus': {
            outline: 'none',
         },
      },

      border: {
         width: '100%',
         height: '100%',
         borderWidth: theme.spacing(1),
         borderStyle: 'dashed',
         borderColor: theme.palette.grey.A100,
         borderRadius: theme.spacing(2),
         padding: theme.spacing(2),

         display: 'flex',
         flexDirection: 'column',
         justifyContent: 'center',
         alignItems: 'center',
         transition: 'border-color 0.5s',
      },

      activeBorder: {
         borderColor: theme.palette.primary.light,
      },

      icon: {
         fontSize: theme.typography.pxToRem(200),
         marginBottom: theme.spacing(1),
         color: theme.palette.grey.A200,
         transition: 'color 0.5s',
      },

      activeIcon: {
         color: theme.palette.primary.light,
      },

      text: {
         ...theme.typography.h4,
         textAlign: 'center',
         color: theme.palette.grey.A400,
      },

      subtext: {
         ...theme.typography.h6,
         textAlign: 'center',
         color: theme.palette.grey.A400,
      },
   })
);

export interface ImageDropProps {
   imageName?: string;
   onImageDropped(files: File[]): void;
}

export const ImageDrop: React.FC<ImageDropProps> = ({ imageName, onImageDropped }) => {
   const classes = useStyles();
   const { getRootProps, getInputProps, isDragActive, isFileDialogActive } = useDropzone({
      onDrop: (files) => onImageDropped(files),
   });

   const isActive = isDragActive || isFileDialogActive;
   const stopPropogation = (event: React.UIEvent): void => event.stopPropagation();

   /* eslint-disable react/jsx-props-no-spreading */
   return (
      <div {...getRootProps()} className={classes.root}>
         <input {...getInputProps()} />
         <div className={`${classes.border} ${isActive ? classes.activeBorder : ''}`}>
            <Panorama
               className={`${classes.icon} ${isActive ? classes.activeIcon : ''}`}
            />
            <div className={classes.text}>
               Just drop {imageName ? `a ${imageName} image` : 'an image'} to instantly
               convert it to PNG
            </div>
            <div className={classes.subtext}>
               Your images are never uploaded (
               <Link
                  href='https://github.com/imgdrop/imgdrop/blob/master/PRIVACY.md'
                  target='_blank'
                  onClick={stopPropogation}
               >
                  Privacy&nbsp;Policy
               </Link>
               )
            </div>
            <div className={classes.subtext}>
               Made with pride by a gay man{' '}
               <span role='img' aria-label='pride flag'>
                  üè≥Ô∏è‚Äçüåà
               </span>{' '}
               (
               <Link
                  href='https://github.com/imgdrop/imgdrop'
                  target='_blank'
                  onClick={stopPropogation}
               >
                  Github
               </Link>
               )
            </div>
         </div>
      </div>
   );
   /* eslint-enable */
};
