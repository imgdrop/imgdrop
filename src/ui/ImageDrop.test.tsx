import React from 'react';
import * as dropzone from 'react-dropzone';
import TestRenderer from 'react-test-renderer';
import { ImageDrop } from './ImageDrop';
import { Link } from '@material-ui/core';

describe(ImageDrop, () => {
   let useDropzoneSpy: jest.SpyInstance;
   let imageDroppedMock: jest.Mock;

   const createRenderer = (imageName?: string): TestRenderer.ReactTestRenderer => {
      return TestRenderer.create(
         <ImageDrop imageName={imageName} onImageDropped={imageDroppedMock} />
      );
   };

   beforeEach(() => {
      useDropzoneSpy = jest.spyOn(dropzone, 'useDropzone');
      imageDroppedMock = jest.fn();
   });

   it('matches the snapshot', () => {
      const renderer = createRenderer();
      expect(renderer.toJSON()).toMatchSnapshot();
   });

   it('matches the snapshot when provided a name', () => {
      const renderer = createRenderer('PNG');
      expect(renderer.toJSON()).toMatchSnapshot();
   });

   it('matches the snapshot when active', () => {
      let setDropzoneState: React.Dispatch<React.SetStateAction<dropzone.DropzoneState>>;
      useDropzoneSpy.mockImplementation(() => {
         let dropzoneState;
         [dropzoneState, setDropzoneState] = React.useState<dropzone.DropzoneState>({
            getRootProps: (): {} => ({}),
            getInputProps: (): {} => ({}),
            isDragActive: false,
            isFileDialogActive: false,
         } as any);
         return dropzoneState;
      });
      const renderer = createRenderer();

      const json = renderer.toJSON();
      TestRenderer.act(() => {
         setDropzoneState((state) => ({
            ...state,
            isDragActive: true,
         }));
      });

      const activeJson = renderer.toJSON();
      expect(activeJson).not.toEqual(json);
      expect(activeJson).toMatchSnapshot();

      TestRenderer.act(() => {
         setDropzoneState((state) => ({
            ...state,
            isDragActive: false,
            isFileDialogActive: true,
         }));
      });

      expect(JSON.stringify(renderer.toJSON())).toBe(JSON.stringify(activeJson));
   });

   it('calls the provided callback when a file is dropped', () => {
      createRenderer();
      expect(useDropzoneSpy).toHaveBeenCalledWith({
         onDrop: expect.any(Function),
      });

      useDropzoneSpy.mock.calls[0][0].onDrop(['file', 'image']);
      expect(imageDroppedMock).toHaveBeenCalledWith(['file', 'image']);
   });

   it('stops propogation on link clicks', () => {
      const renderer = createRenderer();
      const eventMock = {
         stopPropagation: jest.fn()
      };
      renderer.root.findAllByType(Link).forEach(link => {
         link.props.onClick(eventMock);
      });
      expect(eventMock.stopPropagation).toHaveBeenCalledWith();
      expect(eventMock.stopPropagation).toHaveBeenCalledTimes(2);
   });
});
