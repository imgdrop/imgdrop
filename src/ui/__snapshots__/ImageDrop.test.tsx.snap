// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<div
  className="makeStyles-root-1"
  onBlur={[Function]}
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <input
    autoComplete="off"
    multiple={true}
    onChange={[Function]}
    onClick={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex={-1}
    type="file"
  />
  <div
    className="makeStyles-border-2 "
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root makeStyles-icon-4 "
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M23 18V6c0-1.1-.9-2-2-2H3c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h18c1.1 0 2-.9 2-2zM8.5 12.5l2.5 3.01L14.5 11l4.5 6H5l3.5-4.5z"
      />
    </svg>
    <div
      className="makeStyles-text-6"
    >
      Just drop 
      an image
       to instantly convert it to PNG
    </div>
    <div
      className="makeStyles-subtext-7"
    >
      Your images are never uploaded (
      <a
        className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
        href="https://github.com/imgdrop/imgdrop/blob/master/PRIVACY.md"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        target="_blank"
      >
        Privacy Policy
      </a>
      )
    </div>
    <div
      className="makeStyles-subtext-7"
    >
      Made with pride by a gay man
       
      <span
        aria-label="pride flag"
        role="img"
      >
        🏳️‍🌈
      </span>
       
      (
      <a
        className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
        href="https://github.com/imgdrop/imgdrop"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        target="_blank"
      >
        Github
      </a>
      )
    </div>
  </div>
</div>
`;

exports[`matches the snapshot when active 1`] = `
<div
  className="makeStyles-root-1"
>
  <input />
  <div
    className="makeStyles-border-2 makeStyles-activeBorder-3"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root makeStyles-icon-4 makeStyles-activeIcon-5"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M23 18V6c0-1.1-.9-2-2-2H3c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h18c1.1 0 2-.9 2-2zM8.5 12.5l2.5 3.01L14.5 11l4.5 6H5l3.5-4.5z"
      />
    </svg>
    <div
      className="makeStyles-text-6"
    >
      Just drop 
      an image
       to instantly convert it to PNG
    </div>
    <div
      className="makeStyles-subtext-7"
    >
      Your images are never uploaded (
      <a
        className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
        href="https://github.com/imgdrop/imgdrop/blob/master/PRIVACY.md"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        target="_blank"
      >
        Privacy Policy
      </a>
      )
    </div>
    <div
      className="makeStyles-subtext-7"
    >
      Made with pride by a gay man
       
      <span
        aria-label="pride flag"
        role="img"
      >
        🏳️‍🌈
      </span>
       
      (
      <a
        className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
        href="https://github.com/imgdrop/imgdrop"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        target="_blank"
      >
        Github
      </a>
      )
    </div>
  </div>
</div>
`;

exports[`matches the snapshot when provided a name 1`] = `
<div
  className="makeStyles-root-1"
  onBlur={[Function]}
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <input
    autoComplete="off"
    multiple={true}
    onChange={[Function]}
    onClick={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex={-1}
    type="file"
  />
  <div
    className="makeStyles-border-2 "
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root makeStyles-icon-4 "
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M23 18V6c0-1.1-.9-2-2-2H3c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h18c1.1 0 2-.9 2-2zM8.5 12.5l2.5 3.01L14.5 11l4.5 6H5l3.5-4.5z"
      />
    </svg>
    <div
      className="makeStyles-text-6"
    >
      Just drop 
      a PNG image
       to instantly convert it to PNG
    </div>
    <div
      className="makeStyles-subtext-7"
    >
      Your images are never uploaded (
      <a
        className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
        href="https://github.com/imgdrop/imgdrop/blob/master/PRIVACY.md"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        target="_blank"
      >
        Privacy Policy
      </a>
      )
    </div>
    <div
      className="makeStyles-subtext-7"
    >
      Made with pride by a gay man
       
      <span
        aria-label="pride flag"
        role="img"
      >
        🏳️‍🌈
      </span>
       
      (
      <a
        className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
        href="https://github.com/imgdrop/imgdrop"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        target="_blank"
      >
        Github
      </a>
      )
    </div>
  </div>
</div>
`;
